db/migrate/20240206085104_remove_enums.rb

```diff
@@ -0,0 +1,73 @@
+class RemoveEnums < ActiveRecord::Migration[7.1]
+  def up
+    change_column(:storages, :health_status, :string, default: 'pending')
+    add_check_constraint(:storages,
+                         'health_status IN (\'pending\', \'healthy\', \'unhealthy\')',
+                         name: 'storages_health_status_check')
+
+    change_column(:last_project_folders, :mode, :string, default: 'inactive')
+    add_check_constraint(:last_project_folders,
+                         'mode IN (\'inactive\', \'manual\', \'automatic\')',
+                         name: 'last_project_folders_mode_check')
+
+    change_column(:project_storages, :project_folder_mode, :string)
+    add_check_constraint(:project_storages,
+                         'project_folder_mode IN (\'inactive\', \'manual\', \'automatic\')',
+                         name: 'project_storages_project_folder_mode_check')
+
+    change_column(:delayed_job_statuses, :status, :string, default: 'in_queue')
+    add_check_constraint(:delayed_job_statuses,
+                         'status IS NULL OR status IN (\'in_queue\', \'error\', \'in_process\', \'success\', \'failure\', \'cancelled\')',
+                         name: 'delayed_job_statuses_status_check')
+
+    execute <<~SQL.squish
+      DROP TYPE public.delayed_job_status RESTRICT;
+      DROP TYPE public.project_folder_modes RESTRICT;
+      DROP TYPE public.storage_health_statuses RESTRICT;
+    SQL
+  end
+
+  def down
+    execute <<~SQL.squish
+      CREATE TYPE public.delayed_job_status AS ENUM (
+          'in_queue',
+          'error',
+          'in_process',
+          'success',
+          'failure',
+          'cancelled'
+      );
+
+      CREATE TYPE public.project_folder_modes AS ENUM (
+          'inactive',
+          'manual',
+          'automatic'
+      );
+
+      CREATE TYPE public.storage_health_statuses AS ENUM (
+          'pending',
+          'healthy',
+          'unhealthy'
+      );
+    SQL
+
+    remove_check_constraint(:storages, name: 'storages_health_status_check')
+    change_column(:storages, :health_status, :storage_health_statuses, default: nil,
```

Comment:
❓ `default: nil` necessary here for any reason? because the default should be same as in the `up`, just that in the `down` we need to select the default from the enum type, which usually cannot be nil, right?

Reply:
Yes, we need to switch to `NULL` and then in the next step switch it back to the enum type again, see next line. Otherwise the migration fails because it cannot convert the existing default (`string`) to the ENUM

Reply:
So this means, on a down migration we loose all values? or can it convert the string values, just not the default?

Reply:
the `using: 'health_status::storage_health_statuses'` makes sure that all values are converted. it just does not work for the default. Maybe there is a way with typecasting in the `CHANGE COLUMN` call for the default. But the migration is run in a transaction, so this is just the easier way :)